tests/test_precompiled.py

```diff
@@ -0,0 +1,459 @@
+
+from app.precompiled import add_notify_tag_to_letter, validate_document
+from tests.pdf_consts import multi_page_pdf, not_pdf, blank_page, one_page_pdf, no_colour
+
+
+def test_precompiled_validation_endpoint_blank_pdf(client, auth_header):
+
+    response = client.post(
+        url_for('precompiled_blueprint.validate_pdf_document'),
+        data=blank_page,
+        headers={
+            'Content-type': 'application/json',
+            **auth_header
+        }
+    )
+
+    assert response.status_code == 200
+    json_data = json.loads(response.get_data())
+    assert json_data['result'] is True
+
+
+def test_precompiled_validation_endpoint_one_page_pdf(client, auth_header):
+
+    response = client.post(
+        url_for('precompiled_blueprint.validate_pdf_document'),
+        data=one_page_pdf,
+        headers={
+            'Content-type': 'application/json',
+            **auth_header
+        }
+    )
+
+    assert response.status_code == 200
+    json_data = json.loads(response.get_data())
+    assert json_data['result'] is False
+
+
+def test_precompiled_validation_endpoint_no_colour_pdf(client, auth_header):
+
+    response = client.post(
+        url_for('precompiled_blueprint.validate_pdf_document'),
+        data=no_colour,
+        headers={
+            'Content-type': 'application/json',
+            **auth_header
+        }
+    )
+
+    assert response.status_code == 200
+    json_data = json.loads(response.get_data())
+    assert json_data['result'] is False
+
+
+def test_add_notify_tag_to_letter(mocker):
+    file_data = base64.b64decode(multi_page_pdf)
+    pdf_original = PyPDF2.PdfFileReader(BytesIO(file_data))
+
+    assert 'NOTIFY' not in pdf_original.getPage(0).extractText()
+
+    pdf_page = add_notify_tag_to_letter(BytesIO(file_data))
+
+    pdf_new = PyPDF2.PdfFileReader(BytesIO(pdf_page.read()))
+
+    assert pdf_new.numPages == pdf_original.numPages
+    assert pdf_new.getPage(0).extractText() != pdf_original.getPage(0).extractText()
+    assert 'NOTIFY' in pdf_new.getPage(0).extractText()
+    assert pdf_new.getPage(1).extractText() == pdf_original.getPage(1).extractText()
+    assert pdf_new.getPage(2).extractText() == pdf_original.getPage(2).extractText()
+    assert pdf_new.getPage(3).extractText() == pdf_original.getPage(3).extractText()
+
+
+def test_add_notify_tag_to_letter_correct_margins(mocker):
+    file_data = base64.b64decode(multi_page_pdf)
+    pdf_original = PyPDF2.PdfFileReader(BytesIO(file_data))
+
+    can = Canvas(None)
+    # mock_canvas = mocker.patch.object(can, 'drawString')
+
+    can.drawString = MagicMock(return_value=3)
+
+    can.mock_canvas = mocker.patch('app.precompiled.canvas.Canvas', return_value=can)
+
+    file_data = base64.b64decode(multi_page_pdf)
+
+    # It fails because we are mocking but by that time the drawString method has been called so just carry on
+    try:
+        add_notify_tag_to_letter(BytesIO(file_data))
+    except Exception:
+        pass
```

Comment:
```
with pytest.raises(SpecificException):
    add_notify_tag_to_letter(BytesIO(file_data))
```

i don't want to get stung by tests not testing stuff if we refactor the code underneath